Automatically set trust_input=True when function has no inputs

Aarsh-Wankar opened the pull request:
<!-- !! Thank your for opening a PR !! -->

<!--- Provide a self-contained summary of your changes in the Title above -->
<!--- This is what will be shown in the automatic release notes: https://github.com/pymc-devs/pytensor/releases -->

## Description
<!--- Describe your changes in detail -->
An extra `if` condition is added to the `__call__`  function of the `Function` class to check if when `pytorch.function` is called,  the length of the input list is zero. In that case, `trust_input` variable is set to True, thus preventing the execution of a few loops, saving time.


### Tests:
Test code:
```python
import pytensor
import pytensor.tensor as pt
import timeit
fn = pytensor.function([], pt.zeros((5)))
print(timeit.timeit(fn, number=100000))
# time before adding condition:  2.59 µs ± 238 ns per loop (mean ± std. dev. of 10 runs, 100000 loops each)
# time after adding condition:  1.95 µs ± 143 ns per loop (mean ± std. dev. of 10 runs, 100000 loops each)
fn.trust_input=True
print(timeit.timeit(fn, number=100000))
# time before adding condition:  2.29 µs ± 304 ns per loop (mean ± std. dev. of 10 runs, 100000 loops each)
# time after adding condition:  2.04 µs ± 173 ns per loop (mean ± std. dev. of 10 runs, 100000 loops each)
```

## Related Issue
<!--- It is good practice to first open an issue explaining the bug / new feature that is addressed by this PR -->
<!--- Please type an `x` in one of the boxes below and provide the issue number after the # sign: -->
- [x] Closes #1117 
- [ ] Related to #

## Checklist
<!--- Make sure you have completed the following steps before submitting your PR -->
<!--- Feel free to type an `x` in all the boxes below to let us know you have completed the steps: -->
- [x] Checked that [the pre-commit linting/style checks pass](https://docs.pymc.io/en/latest/contributing/python_style.html)
- [x] Included tests that prove the fix is effective or that the new feature works
- [x] Added necessary documentation (docstrings and/or example notebooks)
- [ ] If you are a pro: each commit corresponds to a [relevant logical change](https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes)
<!--- You may find this guide helpful: https://mainmatter.com/blog/2021/05/26/keeping-a-clean-git-history/ -->

## Type of change
<!--- Select one of the categories below by typing an `x` in the box -->
- [x] New feature / enhancement
- [ ] Bug fix
- [ ] Documentation
- [ ] Maintenance
- [ ] Other (please specify):
<!--- Additionally, if you are a maintainer or reviewer, please make sure that the appropriate labels are added to this PR -->

ricardoV94 commented on pytensor/compile/function/types.py:
```diff
@@ -878,6 +878,8 @@ def __call__(self, *args, output_subset=None, **kwargs):
                 output_subset = [self.output_keys.index(key) for key in output_subset]
 
         # Reinitialize each container's 'provided' counter
+        if len(args) == 0:  # for speed we trust the input for empty args
+            trust_input = True
```
This should be done once in the `__init__`, not in every `__call__`

Aarsh-Wankar replied:
Oh, right. I'll change it. Also, yes, I'll take on #1116 as well. Can I make both changes with the same commit? In that case, will these changes work?
```diff
class Function:
    ...
    def __init__(
        self,
        vm: "VM",
        input_storage: list[Container],
        output_storage: list[Container],
        indices,
        outputs,
        defaults,
        unpack_single: bool,
        return_none: bool,
        output_keys,
        maker: "FunctionMaker",
+       trust_input: bool = False,
        name: str | None = None,
    ):
        """
        Parameters
        ----------
        ...
        maker
            The `FunctionMaker` that created this instance.
+       trust_input
+           A boolean variable which indicates whether or not to perform checks on the input. If true, we do not check the input parameter
        name
            A string name.
        """
        self.vm = vm
        self.input_storage = input_storage
        self.output_storage = output_storage
        self.indices = indices
        self.outputs = outputs
        self.defaults = defaults
        self.unpack_single = unpack_single
        self.return_none = return_none
        self.maker = maker
        self.profile = None  # reassigned in FunctionMaker.create
-       self.trust_input = False  # If True, we don't check the input parameter
+      self.trust_input = trust_input # If True, we don't check the input parameter
        self.name = name
        self.nodes_with_inner_function = []
        self.output_keys = output_keys

        if self.output_keys is not None:
            warnings.warn("output_keys is deprecated.", FutureWarning)

        assert len(self.input_storage) == len(self.maker.fgraph.inputs)
        assert len(self.output_storage) == len(self.maker.fgraph.outputs)
+      if len(self.input_storage) == 0:
+            self.trust_input = True # trust the input in case the input is empty for speed.
        self.has_defaults = any(refeed for _, refeed, _ in self.defaults)
```
I will remove the following block from the `__call__` function (from my previous commit).
```python
if len(args) == 0:  # for speed we trust the input for empty args
        trust_input = True
```

ricardoV94 replied:
You have to check if share variables don't count as inputs here since they don't matter for our consideration of trust input. Check the docs for shared variables.

Also we would want a test that it is indeed set to true when we want and not otherwise. 

Yes it can be a single commit since changes are related, but can also be two if you prefer. Should also test the passing trust input directly

ricardoV94 commented:
While you're at it do you want to tackle #1116 as well?

Aarsh-Wankar commented:
@ricardoV94 I added the extra argument trust_input in `pytensor.function` and the intermediate functions too, all the way to the arguments of `Function` class. I check if the length of input (the one given as an input by the user) is zero in `pytensor.function`, and set trust_input to True in that case. Shared variables are added to the input list during the execution of `pfunc`, so there are no shared variables in the initial input list passed to `pytensor.function`.
### Test code
```python
import pytensor
import pytensor.tensor as pt
import timeit
import numpy as np
fn = pytensor.function([], pt.zeros((5)), trust_input=False)
# trust_input is still set to True, as input list is empty.
ar1 = []
ar2 = []
for i in range(10):
    ar1.append(timeit.timeit(fn, number=100000))
print(np.array(ar1).mean(), np.array(ar1).std())
# Output: 0.18654848000151106 0.0037660645585986434

fn.trust_input=False

for i in range(10):
    ar2.append(timeit.timeit(fn, number=100000))
print(np.array(ar2).mean(), np.array(ar2).std()) 
# Output: 0.22520774999284185 0.011584992775607648
```
This also closes #1116.

Are any more changes required?

ricardoV94 commented:
@Aarsh-Wankar sounds like you pulled the main branch with merge (instead of rebase or --ff) which makes it hard to see only the differences introduced by this PR: https://github.com/pymc-devs/pytensor/pull/1206/files

Con you rebase from main instead (and then force push here). If you are unsure, do checkout your branch into a backup branch so you can easily restore it if things go south.

Aarsh-Wankar commented:
@ricardoV94  Sorry for the mistake; this is the first time I am contributing to an open-source repository 😅. Also, thanks for your suggestion, I rebased it to main and force pushed the changes. There are only four commits now. Please let me know if anything needs to be changed.

ricardoV94 commented on pytensor/compile/function/types.py:
```diff
@@ -1303,7 +1308,10 @@ class FunctionMaker:
         print the time spent in this function.
-
+    trust_input : bool
+        If True, the inputs are trusted to be correct. This is used to avoid
+        the overhead of checking the inputs for correctness. This should only
+        be used if the inputs are guaranteed to be correct.
```
Copy the final docstring after adjusting to my comment above.

ricardoV94 commented on pytensor/compile/function/types.py:
```diff
@@ -1469,6 +1477,7 @@ def __init__(
         output_keys=None,
         name=None,
         no_fgraph_prep=False,
+        trust_input=False,
```
Why is this hardcoded false?

Aarsh-Wankar replied:
The default value of trust_input given to `FunctionMaker` is False, but the `Maker` is only called internally in the `orig_function` with the `trust_input` provided by the user (or False otherwise). So if the user calls `function` with `trust_input = True`,  this `trust_input` will propagate to all internal functions, including the `__init__` of `FunctionMaker`.
Do you want me to make trust_input a non-default argument?

ricardoV94 replied:
Sorry I got confused and thought that `trust_input=False` was the permissive setting. My bad

ricardoV94 reviewed (commented):
We need a test for the no inputs case and the allowing passing the keyword argument. The easiest way to check the second case is to repurpose one of the existing tests.

Aarsh-Wankar commented:
@ricardoV94 I added the tests, but some checks seem to be failing. Could you please take a look?

ricardoV94 commented:
Pre-commit is failing you should be able to reproduce locally if you install it

Aarsh-Wankar commented:
Yes, I fixed that; apart from that, a few other tests failed because I didn't add the `trust_input` parameter to the `_Maker`  class in the `DebugMode`. I fixed that, and now only three tests fail. All of them seem to be segmentation faults in tests/tensor. I can't make out how my changes affect these files.

ricardoV94 commented:
I made a mistake. `trust_input` has an effect when there are shared inputs, which are implicit inputs. So our logic of checking if len(inputs)==0 is wrong. 

More importantly, it means the original issue didn't make sense. A function without explicit and implicit inputs is not something we actually care about. That would be a constant function. No need to optimize for that case.

Here is an example:
```python
import pytensor

x = pytensor.shared(5.0)
out = x + 1
fn = pytensor.function([], out)
fn.trust_input  # True (after this PR)
```
However the shared inputs live in the `input_storage` which is checked when `trust_input=False`

```python
fn.input_storage  # [<array(5.)>]
```

So for this PR let's just keep the option to set `trust_input` explicitly, but remove the default of True when len(inputs)==0.

Sorry for the trouble!

Aarsh-Wankar commented:
Oh, not at all. It was fun, and I learned a lot! 😀 Sure, I'll remove the default `trust_input` optimization.

ricardoV94 commented:
Thanks @Aarsh-Wankar
